### PR TITLE
Fixes for 1.2.1

### LINKED ISSUE #798: scenario info window isn't centered
Screenshot 2022-09-03 at 11 22 22

### LINKED ISSUE #1049: "this spell will affect nobody!" dialog doesn't appear
example: my hero has amulet that protects from Berserk, human enemy tries to cast Berserk. When you click Berserk in the spell book window, simply nothing happens.

In oh3 spell book is closed and the dialog appears.

The error is here: https://github.com/vcmi/vcmi/blob/3600d09ece165268171d1882f46edd1691342e75/client/windows/CSpellWindow.cpp#L562-L565

`showInfoDialog()` doesn't show the dialog.

### LINKED ISSUE #1403: advmap spell effect is propagated to units
**Describe the bug**
when a hero casts an advmap spell, then it's visible on the units in a battle as well

**To Reproduce**
1. cast Fly
2. fight someone during this turn

**Expected behavior**
unit count bar is blue, Fly icon doesn't appear in unit details

**Actual behavior**
all units have yellow count bar (as if under negative spell) as well as Fly icon is visible in unit details

**Did it work earlier?**
no idea

**Screenshots**
Screenshot 2023-01-10 at 10 40 11

**Version**
 - OS: macOS Intel
 - Version: the latest develop

**Additional context**
Fly doesn't seem to have any effect on the units

### LINKED ISSUE #1891: Random map templates picker should have full alphabetical sorting
Currently the ingame picker which requires extended RMG submod from vcmi-extras apparently sorts alphabetically only within one submod and concats results

### LINKED ISSUE #1892: Random map templates picker should have list of all templates and manipulate sizes
As in title - current picker from vcmi-extras works in opposite way: requires picking size + underground presence first and then shows templates that match the settings. Sizes not supported by selected template should be greyed out until another template picked, and selection updated if new template is incompatible with previously selected size

### LINKED ISSUE #2012: Bug with Petrification and Fire Shield interaction
**Describe the bug**
If a petrified unit stack dies while under the influence of Fire Shield, the dying unit's corpse does not get placed, instead it switches to idle animation.

**Game logs**
[server_log.txt](https://github.com/vcmi/vcmi/files/11242629/server_log.txt)
[VCMI_Client_log.txt](https://github.com/vcmi/vcmi/files/11242630/VCMI_Client_log.txt)
[VCMI_Launcher_log.txt](https://github.com/vcmi/vcmi/files/11242631/VCMI_Launcher_log.txt)
[VCMI_Server_log.txt](https://github.com/vcmi/vcmi/files/11242632/VCMI_Server_log.txt)
[Testing Petrified Sultans (map).zip](https://github.com/vcmi/vcmi/files/11242634/Testing.Petrified.Sultans.map.zip)

Note: the game logs and the video recording are not from the exact same run.

**To Reproduce**
Needs:
- Unit stack with the Petrification ability
- Enemy unit stack with Fire Shield active

Steps:
- Use a unit stack with Fire Shield to attack an enemy unit with the Petrification ability
- Fire Shield unit gets Petrified
- kill the Petrified unit
- experience bug

**Expected behavior**
The petrified unit equipped with Fire Shield should leave behind a corpse after dying.

**Actual behavior**
The petrified unit plays the dying animation, but after the Fire Shield kicks in, it leaves behind an idle animation model instead of a corpse.

**Did it work earlier?**
Unknown.

**Screenshots**
Video link: https://mega.nz/file/dMlxHQ6D#Z5iWqDlrUvRXONrTxdOllK5KgW5lAjpBgD8AyJ3gYds

**Version**
 - OS: Windows 10
 - Version: VCMI 1.2 release version

**Additional context**
Test map has been tested with the following aid from Ivan on Discord
`You can change Stone Gaze chance to 100% to ease testing
config/creatures/dungeon.json
Line 199 for base medusa, line 233 for upgraded:
"val" : 20,
Changing to 100 would give you guaranteed petrification`

### LINKED ISSUE #2014: VCMI Client crash when leaving from hot seat battle, to starting a new hot seat
**Describe the bug**
The VCMI Client crashes when someone returns to the main menu during a hotseat combat, and then tries to start a new hotseat session

**Game logs**
[server_log.txt](https://github.com/vcmi/vcmi/files/11242926/server_log.txt)
[VCMI_Client_log.txt](https://github.com/vcmi/vcmi/files/11242927/VCMI_Client_log.txt)
[VCMI_Launcher_log.txt](https://github.com/vcmi/vcmi/files/11242929/VCMI_Launcher_log.txt)
[VCMI_Server_log.txt](https://github.com/vcmi/vcmi/files/11242930/VCMI_Server_log.txt)

[Testing Hotseat Battle Crash (map).zip](https://github.com/vcmi/vcmi/files/11242932/Testing.Hotseat.Battle.Crash.map.zip)

**To Reproduce**
- Start Launcher
- Start Game
- New Game -> Multiplayer -> Hot Seat
- name the second player
- OK
- select and start test map
- Red player's turn starts
- attack an enemy hero or town (crash reproducible in both cases)
- select Combat Options during combat
- return to Main Menu (crash reproducible with either side in turn)
- New Game -> Multiplayer -> Hot Seat
- attempt to name second player
- VCMI Client crashes

**Expected behavior**
VCMI Client should not crash

**Actual behavior**
VCMI Client crashes

**Did it work earlier?**
Unknown.

**Screenshots**
N/A

**Version**
 - OS: Windows 10
 - Version: VCMI 1.2.0 version

**Additional context**
- Tried finding duplicates, bug #1962 or #1044 stood out to me, but I can't tell if they're really about the same core issue
- First ran into this problem when tuning my test map for bug #2012

### LINKED ISSUE #2016: Crash when filtering saves on save game screen result in no saves
**Describe the bug**
When filtering saves on save game screen, game crashes if there are no saves for selected map size.

**Game logs**
[server_log.txt](https://github.com/vcmi/vcmi/files/11243022/server_log.txt)
[VCMI_Client_log.txt](https://github.com/vcmi/vcmi/files/11243023/VCMI_Client_log.txt)
[VCMI_Server_log.txt](https://github.com/vcmi/vcmi/files/11243024/VCMI_Server_log.txt)

**To Reproduce**
1. Make sure you don't have saves for at least one of map sizes(those that can be selected)
2. Start any map
3. Menu -> save game
4. Select map size filter on the top

**Expected behavior**
Should show empty list.

**Actual behavior**
client crash with
```
ERROR global - Disaster happened.
ERROR global - Reason: 0xc0000005 - EXCEPTION_ACCESS_VIOLATION at 0033:0x13f,aa3,355
ERROR global - Attempt to read from 0x    0x18
ERROR global - Thread ID: 5412
ERROR global - Crash info will be put in VCMI_client.exe_crashinfo.dmp
```

**Did it work earlier?**
no idea

**Screenshots**

**Version**
 - OS: Windows 7
 - Version: 1.2

**Additional context**

### LINKED ISSUE #2026: Black Market seems to be missing the goods
**Describe the bug**
Visiting a Black Market seems to be 

**Game logs**
[server_log.txt](https://github.com/vcmi/vcmi/files/11264809/server_log.txt)
[VCMI_Client_log.txt](https://github.com/vcmi/vcmi/files/11264810/VCMI_Client_log.txt)
[VCMI_Launcher_log.txt](https://github.com/vcmi/vcmi/files/11264811/VCMI_Launcher_log.txt)
[VCMI_Server_log.txt](https://github.com/vcmi/vcmi/files/11264812/VCMI_Server_log.txt)

**To Reproduce**
Visit a Black Market on the map.

**Expected behavior**
I should see the montly list of artifacts up for sale (or at least what is left of them).

**Actual behavior**
The artifacts are missing from the Black Market.

**Did it work earlier?**
Yes, I did not experiences such problems in 1.1.0 or 1.1.1

**Screenshots**
Screenshot 2023-04-18 19-41-34

**Version**
 - OS: Windows 10
 - Version: VCMI 1.2.0 release version

**Additional context**
In-Town markets work properly, only the external Black Market looks buggy.

### LINKED ISSUE #2034: Morale works incorrectly
**Describe the bug**
Noticed this behavior with negative morale. Thus, negative morale debuff is checked twice per turn.

**To Reproduce**
1. Have a negative morale
2. Start battle
3. Wait with your creatures till end of turn

**Expected behavior**
No negative morale at same turn for creatures available to wait earlier

**Actual behavior**
Negative morale appears for some awaited creatures. 

**Did it work earlier?**
Yes, it did

**Version**
 - Version: develop 1.3

### PR DESCRIPTION
**Pull request by IvanSavenko**
Fixes for issues discovered in 1.2 release. To be released as 1.2.1
- Implements #843 
- Fixes #798 
- Fixes #1049
- Fixes #1403 
- Fixes #1891
- Fixes #1892
- Fixes #2012 
- Fixes #2014
- Fixes #2016 
- Fixes #2026 
- Fixes #2034

### CODE REVIEW

**Comment by Nordsoft91** [comment]
@IvanSavenko , could you please cherry-pick https://github.com/vcmi/vcmi/pull/2033?

**Comment by IvanSavenko** [comment]
@Nordsoft91 Understood, will do.

**Comment by IvanSavenko** [comment]
Probably all for now, will recheck few remaining issues in 1.2.1 milestone later.

File: client/lobby/CSavingScreen.cpp [hunk 1]
```diff
@@ -52,7 +50,9 @@ const CMapInfo * CSavingScreen::getMapInfo()
 
 const StartInfo * CSavingScreen::getStartInfo()
 {
-	return localSi;
+	if (localMi)
+		return localMi->scenarioOptionsOfSave;
+	return LOCPLINT->cb->getStartInfo();
```
Comment (Nordsoft91) [inline client/lobby/CSavingScreen.cpp:55]: Are local interface always exist? Suggest to test mp scenarios

File: client/mapView/MapRendererContext.cpp [hunk 2]
```diff
@@ -266,6 +272,16 @@ bool MapRendererAdventureContext::showBlocked() const
 	return settingShowBlocked;
 }
 
+bool MapRendererAdventureContext::showSpellRange(const int3 & position) const
+{
+	if (!settingSpellRange)
+		return false;
+
+	auto hero = adventureInt->curHero();
```
Comment (Nordsoft91) [inline client/mapView/MapRendererContext.cpp:280]: Check that both adventureInt and hero are not nullptr?

File: client/lobby/CSavingScreen.cpp (hunk 1 above)
Comment (IvanSavenko) [reply client/lobby/CSavingScreen.cpp:55]: Yes, always exists.
In MP game switches between different LOCPLINT'S.

Old code was also using LOCPLINT unconditionally.
LOCPLINT is null during pregame, but saving screen is not available in it.

File: client/mapView/MapRendererContext.cpp (hunk 2 above)
Comment (IvanSavenko) [reply client/mapView/MapRendererContext.cpp:280]: adventureInt - always exists
curHero - *should* always exist, will add check just in case

**Pull request merged**